toxcore/query.c

```diff
@@ -0,0 +1,347 @@
+        P_QUERY test = queries->query_list[i];
+
+        if (!q_verify_server(test.ipp, pend.ipp)) {
+            continue;
+        }
+
+        if (!id_equal(test.key, pend.key)) {
+            continue;
+        }
+
+        if (outgoing) {
+            if (test.length != pend.length) {
+                continue;
+            }
+
+            if (memcmp(test.name, pend.name, test.length) != 0) {
+                continue;
+            }
+        }
+
+        // if (memcmp(test.nonce, pend.nonce, ) {
+        // continue;
+        // }
+
+        return i;
+    }
+
+    return -1;
+}
+
+static bool q_add(PENDING_QUERIES *queries, P_QUERY pend)
+{
+    if (queries->count >= queries->size) {
+        if (q_grow(queries) != true) {
+            return false;
+        }
+    }
+
+    queries->query_list[queries->count] = pend;
+    ++queries->count;
+
+    return true;
+}
+
+/** q_find_and_drop
+ *
+ * checks for a response in the pending queries list, if it exists, it'll drop it from the list.
+ */
+static bool q_drop(PENDING_QUERIES *queries, size_t loc)
+{
+    if (loc == -1) {
+        return false;
+    }
+
+    free(queries->query_list[loc].name);
+    --queries->count;
+
+    if (loc == queries->count) {
+        return true;
+    }
+
+    memmove(&queries->query_list[loc], &queries->query_list[loc + 1], sizeof(P_QUERY));
+    return true;
+}
+
+static size_t q_build_pkt(const uint8_t *their_public_key, const uint8_t *our_public_key, const uint8_t *our_secret_key,
+                          uint8_t type, const uint8_t *data, size_t length, uint8_t *built)
+{
+    // Encrypt the outgoing data
+    uint8_t nonce[crypto_box_NONCEBYTES];
+    new_nonce(nonce);
+
+    uint8_t encrypted[length + crypto_box_MACBYTES];
+    int status = encrypt_data(their_public_key, our_secret_key, nonce, data, length, encrypted);
+
+    if (status == -1) {
+        return -1;
+    }
+
+    // Build the packet
+    size_t size = 0;
+    built[0] = type;
+    size += 1;
+
+    memcpy(built + size, our_public_key, crypto_box_PUBLICKEYBYTES);
+    size += crypto_box_PUBLICKEYBYTES;
+
+    memcpy(built + size, nonce, crypto_box_NONCEBYTES);
+    size += crypto_box_NONCEBYTES;
+
+    memcpy(built + size, encrypted, status);
+
+    return 1 + crypto_box_PUBLICKEYBYTES + crypto_box_NONCEBYTES + status;
+}
+
+static int q_send(DHT *dht, P_QUERY send)
+{
+    uint8_t packet[QUERY_PKT_ENCRYPTED_SIZE(send.length)];
+
+    size_t written_size = q_build_pkt(send.key, dht->self_public_key, dht->self_secret_key, NET_PACKET_DATA_REQUEST,
+                                      send.name, send.length, packet);
+    // TODO(grayhatter) add tox_assert(written_size == QUERY_PKT_ENCRYPTED_SIZE(send.length));
+
+    if (written_size != -1) {
+        return sendpacket(dht->net, send.ipp, packet, written_size) != written_size;
+    }
+
+    return -1;
+}
+
+static P_QUERY q_make(IP_Port ipp, const uint8_t key[TOX_PUBLIC_KEY_SIZE], const uint8_t *name, size_t length)
```

Comment:
Why doesn't this return a pointer? and why doesn't it take ipp and key as pointers?